cmd/zrepl/zrepl.c

```diff
@@ -213,33 +217,27 @@ uzfs_zvol_io_receiver(void *arg)
 		taskq_dispatch(zinfo->uzfs_zvol_taskq, uzfs_zvol_worker,
 		    zio_cmd, TQ_SLEEP);
 	}
-exit:
-	if (zinfo != NULL) {
-		LOG_DEBUG("uzfs_zvol_io_receiver thread for zvol %s exiting",
-		    zinfo->name);
-		(void) pthread_mutex_lock(&zinfo->zinfo_mutex);
-		zinfo->conn_closed = B_TRUE;
-		/*
-		 * Send signal to ack sender so that it can free
-		 * zio_cmd, close fd and exit.
-		 */
-		if (zinfo->io_ack_waiting) {
-			rc = pthread_cond_signal(&zinfo->io_ack_cond);
-		}
-		/*
-		 * wait for ack thread to exit to avoid races with new
-		 * connections for the same zinfo
-		 */
-		while (zinfo->conn_closed && zinfo->is_io_ack_sender_created) {
-			(void) pthread_mutex_unlock(&zinfo->zinfo_mutex);
-			usleep(1000);
-			(void) pthread_mutex_lock(&zinfo->zinfo_mutex);
-		}
+
+	(void) pthread_mutex_lock(&zinfo->zinfo_mutex);
+	zinfo->conn_closed = B_TRUE;
+	/*
+	 * Send signal to ack sender so that it can free
+	 * zio_cmd, close fd and exit.
+	 */
+	if (zinfo->io_ack_waiting) {
+		rc = pthread_cond_signal(&zinfo->io_ack_cond);
+	}
+	/*
+	 * wait for ack thread to exit to avoid races with new
+	 * connections for the same zinfo
+	 */
+	while (zinfo->conn_closed && zinfo->is_io_ack_sender_created) {
 		(void) pthread_mutex_unlock(&zinfo->zinfo_mutex);
-		uzfs_zinfo_drop_refcnt(zinfo, B_FALSE);
-	} else {
-		LOG_DEBUG("uzfs_zvol_io_receiver thread exiting");
+		usleep(1000);
+		(void) pthread_mutex_lock(&zinfo->zinfo_mutex);
 	}
+	(void) pthread_mutex_unlock(&zinfo->zinfo_mutex);
+	uzfs_zinfo_drop_refcnt(zinfo, B_FALSE);
```

Comment:
IMO, we should log that io_receiver thread is exiting for xxxx volume.

Reply:
There is another log message with similar meaning and more understandable. When io ack thread exits it prints `"Data connection for zvol %s closed"`. This info msg eliminates need for the debug message here because this thread exits just after the io ack thread exits and prints this message.

In my opinion in general we should avoid messages referring to thread names, file descriptor numbers, etc. and prefer messages which refer to replication protocol events instead. Those require lesser knowledge of implementation details and are easier to understand.